rancher2/resource_rancher2_cluster.go

```diff
@@ -385,93 +385,57 @@ func resourceRancher2ClusterUpdate(d *schema.ResourceData, meta interface{}) err
-	}
-
-	stateConf := &resource.StateChangeConf{
-		Pending:    []string{"active", "provisioning", "pending", "updating", "upgrading"},
-		Target:     []string{"active", "provisioning", "pending"},
-		Refresh:    clusterStateRefreshFunc(client, newCluster.ID),
-		Timeout:    d.Timeout(schema.TimeoutUpdate),
-		Delay:      1 * time.Second,
-		MinTimeout: 3 * time.Second,
-	}
-	_, waitErr := stateConf.WaitForState()
-	if waitErr != nil {
-		return fmt.Errorf("[ERROR] waiting for cluster (%s) to be updated: %s", newCluster.ID, waitErr)
-	}
-
-	// cluster_monitoring is updated here
-	if d.HasChange("enable_cluster_monitoring") || d.HasChange("cluster_monitoring_input") {
-		clusterResource := &norman.Resource{
-			ID:      newCluster.ID,
-			Type:    newCluster.Type,
-			Links:   newCluster.Links,
-			Actions: newCluster.Actions,
+	// update the cluster; retry til timeout or non retryable error is returned. If api 500 error is received,
+	// retry to see if update will go through
+	return resource.Retry(d.Timeout(schema.TimeoutUpdate), func() *resource.RetryError {
+		newCluster := &Cluster{}
+		if replace {
+			err = client.APIBaseClient.Replace(managementClient.ClusterType, cluster, update, newCluster)
+		} else {
+			err = client.APIBaseClient.Update(managementClient.ClusterType, cluster, update, newCluster)
 		}
-		enableMonitoring := d.Get("enable_cluster_monitoring").(bool)
-		if enableMonitoring {
-			monitoringInput := expandMonitoringInput(d.Get("cluster_monitoring_input").([]interface{}))
-			if len(newCluster.Actions[monitoringActionEnable]) > 0 {
-				err = client.APIBaseClient.Action(managementClient.ClusterType, monitoringActionEnable, clusterResource, monitoringInput, nil)
-				if err != nil {
-					return err
-				}
-			} else {
-				monitorVersionChanged := false
-				if d.HasChange("cluster_monitoring_input") {
-					old, new := d.GetChange("cluster_monitoring_input")
-					oldInput := old.([]interface{})
-					oldInputLen := len(oldInput)
-					oldVersion := ""
-					if oldInputLen > 0 {
-						oldRow, oldOK := oldInput[0].(map[string]interface{})
-						if oldOK {
-							oldVersion = oldRow["version"].(string)
-						}
-					}
-					newInput := new.([]interface{})
-					newInputLen := len(newInput)
-					newVersion := ""
-					if newInputLen > 0 {
-						newRow, newOK := newInput[0].(map[string]interface{})
-						if newOK {
-							newVersion = newRow["version"].(string)
-						}
-					}
-					if oldVersion != newVersion {
-						monitorVersionChanged = true
-					}
-				}
-				if monitorVersionChanged && monitoringInput != nil {
-					err = updateClusterMonitoringApps(meta, d.Get("system_project_id").(string), monitoringInput.Version)
-					if err != nil {
-						return err
-					}
-				}
-				err = client.APIBaseClient.Action(managementClient.ClusterType, monitoringActionEdit, clusterResource, monitoringInput, nil)
-				if err != nil {
-					return err
-				}
+		if err != nil {
+			if IsServerError(err) {
+				return nil
```

Comment:
Q: Should we return the error or warp it as a RetryableError here so that a retry will be kicked off? 
Sorry if I misunderstand the code here.

Reply:
Nil returned _will_ retry the error. Anything except for a `resource.Retry` timeout or a `NonRetryableError` will continue to retry the API call.

Reply:
The `resource.Retry` method will return success if no error is returned from the argument function. If we want to retry on server errors, the argument function should return a `RetryableError` instead of `nil`.

Reply:
You, are right. Good catch